### DISCUSSION TITLE
Virtual Hosting with Axum

### DISCUSSION THREAD

**Discussion by apps4uco** [tokio-rs/axum#2872] [Q&A] [answered]
### Summary

Hi, I want to do virtual hosting with Axum and I found https://github.com/skerkour/kerkour.com/tree/main/blog/2022/rust_axum_hostname_router/src which works with the latest Axum however I understand that the recommended way of adding state is with with_state. When I try this it fails to compile with the message:

```
error[E0599]: the method `oneshot` exists for struct `Router<Arc<AppState>>`, but its trait bounds were not satisfied
  --> src/main.rs:34:41
   |
34 |                     _ => website_router.oneshot(request).await,
   |                                         ^^^^^^^
   |
   |
62 | pub struct Router<S = ()> {
   | ------------------------- doesn't satisfy `Router<Arc<AppState>>: Service<_>` or `Router<Arc<AppState>>: tower::ServiceExt<_>`
   |
   = note: the following trait bounds were not satisfied:
           `Router<Arc<AppState>>: Service<_>`
           which is required by `Router<Arc<AppState>>: tower::ServiceExt<_>`
           `&Router<Arc<AppState>>: Service<_>`
           which is required by `&Router<Arc<AppState>>: tower::ServiceExt<_>`
           `&mut Router<Arc<AppState>>: Service<_>`
           which is required by `&mut Router<Arc<AppState>>: tower::ServiceExt<_>`

```

Code here

[main.rs.txt](https://github.com/user-attachments/files/16592196/main.rs.txt)

Can anyone help me solve this?

Also it seems like a common enough requirement that it would be great to add as an example.

By the way the original example doesnt work completely correctly as the host extractor includes the port as well.

### axum version

0.7.5

**Comment by jplatte** [comment, marked as answer]
You're supplying the state at the wrong place. `api_router` and `website_router`'s handler need access to it, not `app`. When using `nest`, the required state type is automatically propagated to the parent router (if there are no conflicting requirements), but that only works for `nest`, not this `oneshot`-based router nesting that you're doing.

**Reply by apps4uco** [reply]
Great, thanks for your reply, moving the state as you said to both routes compiles.
```
let api_router = Router::new()
        .route("/*path", any(api_handler))
        .with_state(state.clone());

    let website_router = Router::new()
        .route("/*path", any(website_handler))
        .with_state(state);
```

I wonder if you could help with one more thing, there is still a problem, the root path "/" is not matched by "/*path" and also if the handlers also have a path component, then the handler has to have 2 paths one for the "dummy" path used to match the host and one for the real path.

```
async fn website_handler(
    Path((_every_path, path)): Path<(String, String)>,
    State(state): State<Arc<AppState>>,
) -> Html<String> {
```

Whereas what I would like to have is  
```
async fn website_handler(
    Path(path): Path<String>,
    State(state): State<Arc<AppState>>,
) -> Html<String> {

```

What would be the best way to solve both these problems? By the way, If you prefer I can mark your original answer as the solution and start a different discussion.

New Code: 
[main.rs.txt](https://github.com/user-attachments/files/16594264/main.rs.txt)

**Comment by apps4uco** [comment]
Thanks to @jplatte I figured out one way to create virtual hosts in Axum.

In Cargo.toml add the steer feature of tower

```
tower = {version="0.4",features=["steer"]}
```

Define routers for each virtual host, and for each one call with_state, as shown in previous comment.

Define the virtual host service that selects which router to use based on the host header, and/or forwarded headers (not shown)

```
let mut virtual_host_service = Steer::new(
        // All virtual host routers converted to services
        vec![website_router.into_service(), api_router.into_service()],
        // Pick which virtual host service to send the request to
        |req: &Request<Body>, _services: &[_]| {
            let host = req
                .headers()
                .get(http::header::HOST)
                .and_then(|host| host.to_str().ok());

            let hostname_port = host.map(|host| {
                host.split_once(':')
                    .map(|(hostname, port_str)| (hostname, port_str.parse::<u16>().unwrap()))
                    .unwrap_or((&host, 80u16))
            });

            let hostname = hostname_port.map(|(hostname, _port)| hostname);

            // match the hostname to the index in the virtual host service vector
            match hostname {
                // put your hostnames here
                Some("apps4u.local") => 1,
                //default host if none match, ie in testing localhost 127.0.0.1
                _ => 0,
            }
        },
    );
```
This code could be more sophisticated by defining a struct for each Virtual host, that make sure you cant get the indices mixed up, by for example inserting more virtual hosts or changing the order in the vec. However, you probably want to do some fairly custom code for each virtual host like matching on subdomains or accepting www.example.com and example.com anyway, in any case this code is more of a how to than production code. 

Use the router, note any common routes can be defined here.

(In the test case I had the type of the router had to be defined as Router<()> however it may not always be necessary.)

```
let app: Router<()> = Router::new().fallback_service(virtual_host_service);
```
